Update/set project up with cookie cutter template

NicholasBunn opened the pull request:
This PR reshapes the forked project to fit into the [Kraken cookie cutter repository structure](https://github.com/octoenergy/cookiecutter-kraken-package). This allows it to make use of all the tooling already on offer there and standardises it with other open source projects currently being maintained by Kraken

Nox results
![Screenshot 2024-08-20 at 10 52 10](https://github.com/user-attachments/assets/728b40cf-e2fa-4218-80f9-2482d88edf41)

NicholasBunn committed:
Move files into expected dirs

Move the actual package code into a src directory following the Kraken open source package convention, update the test directory structure to support our test tooling

NicholasBunn committed:
Add makefile

NicholasBunn committed:
Add pre-commit setup file

NicholasBunn committed:
Update package requirements

Replace the existing requirements file with auto-generated ones and the necessary pyproject.toml to generate the files

NicholasBunn committed:
Update git ignore

NicholasBunn committed:
Update README

NicholasBunn committed:
Swap Tox for Nox

Switched the testing framework from Tox to Nox. In doing this, we've dropped support for DRF 3.12 because I'm still getting my head around Nox and want to get this PR out. We've also removed the travis setup here as we'll be using CircleCI instead."

NicholasBunn committed:
Remove manifest

NicholasBunn committed:
Add changelog

NicholasBunn commented on makefile:
🍷 This is just the standard makefile from the cookie-cutter template

NicholasBunn commented on .pre-commit-config.yaml:
🍷 This is just the standard pre-commit config from the cookie-cutter template, no changes have been made

NicholasBunn commented on pyproject.toml:
🍷 This is just the standard pyproject file from the cookie-cutter template. The only updates that have been made is the inclusion of this project's dependencies, highlighted below.

NicholasBunn commented on pyproject.toml:
```diff
@@ -0,0 +1,168 @@
+# Do not manually edit the version, use `make version_{type}` instead.
+# This should match the version in the [tool.bumpversion] section.
+version = "0.1.0"
+dependencies = [
+    "Django>=3.2",
+    "djangorestframework>=3.12.0",
+    "pytest-django==4.7.0",
+    "pytest==8.3.2",
+    "pytest-cov",
+]
```
🍷 These were added in for this package, specifically. They were taken from the existing `requirements.txt`

NicholasBunn commented on README.md:
🍷 This is just the standard README from the cookie-cutter template, updated to be relevant to this package

NicholasBunn commented on noxfile.py:
```diff
@@ -0,0 +1,88 @@
+@nox.session()
+@nox.parametrize(
+    "drf_version",
+    [
+        nox.param("djangorestframework>=3.13,<3.14", id="drf=3.13"),
+        nox.param("djangorestframework>=3.14,<3.15", id="drf=3.14"),
+        nox.param("djangorestframework>=3.15,<3.16", id="drf=3.15"),
+    ],
```
🍷 Note here that we've dropped test support for 3.12. This is just to get this parametrization working as 3.12 is not supported beyond Django 3.2 and I didn't want to spend too much time playing around with the Nox file if there is nobody actively using that combination of this package yet

NicholasBunn commented on noxfile.py:
🍷 Apart form the parametrization, this is just the standard noxfile from the cookie-cutter template

NicholasBunn commented on MANIFEST.in:
🍷 Don't think this is needed in the updated configurations

isabellanorris reviewed (approved):
All looks reasonable to be, might be worth some guild eyes on this at some point

NicholasBunn commented on .pre-commit-config.yaml:
🍷 This is just the standard pre-commit config from the cookie-cutter template, no changes have been made

NicholasBunn commented on makefile:
🍷 This is just the standard makefile from the cookie-cutter template, no changes have been made

NicholasBunn commented on README.md:
🍷 This is just the standard README from the cookie-cutter template, updated to be relevant to this package

NicholasBunn commented on pyproject.toml:
```diff
@@ -0,0 +1,168 @@
+# Do not manually edit the version, use `make version_{type}` instead.
+# This should match the version in the [tool.bumpversion] section.
+version = "0.1.0"
+dependencies = [
+    "Django>=3.2",
+    "djangorestframework>=3.13.0",
+    "pytest-django==4.7.0",
+    "pytest==8.3.2",
+    "pytest-cov",
+]
```
🍷 These were added in for this package, specifically. They were taken from the existing requirements.txt

NicholasBunn replied:
Note also that we're bumping the min `djangorestframework` version from **3.12** -> **3.13** here. 3.12 still works, I just needed a quick workaround to get the `Nox` file passing so that we unblock work requiring this package

NicholasBunn commented on MANIFEST.in:
🍷 Don't think this is needed in the updated configurations

NicholasBunn committed:
Add CircleCI config

Note that we've disabled MyPy here, this is because mypy is currently failing due to this project being untyped. I want to set CircleCI up to run after merging this PR and don't want it to fail on master. I plan to add typing and re-enable MyPy in a subsequent PR

NicholasBunn commented on .circleci/config.yml:
```diff
@@ -0,0 +1,87 @@
+          name: Run ruff linter
+          command: make ruff_lint
+          when: always
+      # - run:
+      #     name: Run Mypy
+      #     command: make mypy
+      #     when: always
```
🍷 Note that we've disabled MyPy here to avoid failing CI once this branch is merged. This will be re-enabled in the PR where we add type hints to this repo

NicholasBunn committed:
Format CODEOWNERS

Just so that pre-commit check doesn't fail on CI

NicholasBunn commented on .circleci/config.yml:
```diff
@@ -0,0 +1,87 @@
+      - checkout
+      - run:
+          name: Install additional Python versions and system Python packages
+          command: |
+            # Install additional Python versions that Nox needs. Note the
+            # `cimg/python:3.11` image already has Python 3.10 installed.
+            pyenv install 3.12.3
+            pyenv global 3.11.9 3.12.3
+            pipx install uv nox
+          when: always
```
❓ Because Python 3.10 is included in `cimg/python:3.11`, we shouldn't need to install it here, should we?

delfick replied:
the `pyenv global` is only saying what to use as the default and we only have `pyenv install` for 3.12

delfick reviewed (approved):
cool
